Combine javascript and javascript-apollo generators

joaomlneto opened the pull request:
This PR merges the functionality of the `javascript-apollo` into the `javascript` generator.
Target generation is specified via the `library` command line option, similar to the `spring` generator.

`openapi-generator generate -g javascript-apollo` ➡️  `openapi-generator generate -g javascript --library apollo`

---

CC Node/Javascript technical committee: @CodeNinjai @frol @cliffano 

<!-- Please check the completed items below -->
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.1.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

wing328 commented:
@joaomlneto thanks for the PR. Can you please run `./bin/utils/ensure-up-to-date` to update the samples and the doc?

joaomlneto commented:
Everything should be OK now! Sorry I'm a bit of an idiot, and not used to codebases this big:
(1) Although the IDE was configured to use Java 8, the scripts were using Java 17.
(2) Afterwards I forgot to pull my very own changes from upstream, which were causing the generated samples to differ.

Note to future self:
```sh
export JAVA_HOME=`/usr/libexec/java_home -v 1.8
# or
./run-in-docker.sh mvn clean package && ./run-in-docker.sh /gen/bin/generate-samples.sh && ./run-in-docker.sh /gen/bin/utils/export_docs_generators.sh
```

wing328 commented:
When testing javascript-apollo petstore, I got the following:
```
➜  javascript-apollo git:(joaomlneto-combine-apollo-generator) ✗ npm install
npm WARN deprecated formidable@1.2.6: Please upgrade to latest, formidable@v2 or formidable@v3! Check these notes: https://bit.ly/2ZEqIau
npm WARN deprecated superagent@5.3.1: Please upgrade to v7.0.2+ of superagent.  We have fixed numerous issues with streams, form-data, attach(), filesystem errors not bubbling up (ENOENT on attach()), and all tests are now passing.  See the releases tab for more information at <https://github.com/visionmedia/superagent/releases>.

> open_api_petstore@1.0.0 prepare
> npm run build


> open_api_petstore@1.0.0 build
> babel src -d dist

SyntaxError: /Users/williamcheng/Code/openapi-generator5/samples/client/petstore/javascript-apollo/src/model/List.js: Unexpected token (26:7)

  24 |      * @type {String}
  25 |      */
> 26 |     123-list;
     |        ^
  27 |
  28 |     
  29 |
    at instantiate (/Users/williamcheng/Code/openapi-generator5/samples/client/petstore/javascript-apollo/node_modules/@babel/parser/lib/index.js:72:32)
    at constructor (/Users/williamcheng/Code/openapi-generator5/samples/client/petstore/javascript-apollo/node_modules/@babel/parser/lib/index.js:366:12)
    at Parser.raise (/Users/williamcheng/Code/openapi-generator5/samples/client/petstore/javascript-apollo/node_modules/@babel/parser/lib/index.js:3453:19)
    at Parser.unexpected (/Users/williamcheng/Code/openapi-generator5/samples/client/petstore/javascript-apollo/node_modules/@babel/parser/lib/index.js:3491:16)
    at Parser.parseClassMemberWithIsStatic (/Users/williamcheng/Code/openapi-generator5/samples/client/petstore/javascript-apollo/node_modules/@babel/parser/lib/index.js:15850:12)
    at Parser.parseClassMember (/Users/williamcheng/Code/openapi-generator5/samples/client/petstore/javascript-apollo/node_modules/@babel/parser/lib/index.js:15717:10)
    at /Users/williamcheng/Code/openapi-generator5/samples/client/petstore/javascript-apollo/node_modules/@babel/parser/lib/index.js:15657:14
    at Parser.withSmartMixTopicForbiddingContext (/Users/williamcheng/Code/openapi-generator5/samples/client/petstore/javascript-apollo/node_modules/@babel/parser/lib/index.js:14457:14)
    at Parser.parseClassBody (/Users/williamcheng/Code/openapi-generator5/samples/client/petstore/javascript-apollo/node_modules/@babel/parser/lib/index.js:15632:10)
    at Parser.parseClass (/Users/williamcheng/Code/openapi-generator5/samples/client/petstore/javascript-apollo/node_modules/@babel/parser/lib/index.js:15606:22) {
  code: 'BABEL_PARSE_ERROR',
  reasonCode: 'UnexpectedToken',
  loc: Position { line: 26, column: 7, index: 614 },
  pos: [Getter/Setter]
}
npm ERR! code 1
npm ERR! path /Users/williamcheng/Code/openapi-generator5/samples/client/petstore/javascript-apollo
npm ERR! command failed
npm ERR! command sh -c npm run build

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/williamcheng/.npm/_logs/2022-08-25T02_24_18_717Z-debug.log

```

wing328 commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavascriptApolloClientCodegen.java:
```diff
@@ -104,6 +104,10 @@ public JavascriptApolloClientCodegen() {
+        generatorMetadata = GeneratorMetadata.newBuilder(generatorMetadata)
+                .stability(Stability.DEPRECATED)
+                .build();
+
```
In this file, can you please also rename `javascript-apollo` to `javascript-apollo-deprecated`?

wing328 commented:
cc @erithmetic (creator of javascript-apollo)

joaomlneto commented:
@wing328 I've fixed the generated code compilation via https://github.com/OpenAPITools/openapi-generator/pull/13270/commits/05d8e396109856bbaeff0b0eeb59f2e3f7e96fba

I've also renamed the `javascript-apollo` generator to `javascript-apollo-deprecated` via https://github.com/OpenAPITools/openapi-generator/pull/13270/commits/3e44113257a1bf7b9e42739ddb51744c09d2afcc, if I understood you correctly.

I've ran into an issue where running `./run-in-docker.sh /gen/bin/utils/export_docs_generators.sh` wouldn't update the `docs/generators.md` file 🤷🏻‍♂️ .
I've decided to sin and manually edit the file via https://github.com/OpenAPITools/openapi-generator/pull/13270/commits/f7f85ddcde5fac6be439d65718b4110ecec20dde

wing328 commented:
@joaomlneto thanks for the prompt fix.

I can now run `npm install` without issues but running `npm test` returns the following issues:
```
➜  javascript-apollo git:(joaomlneto-combine-apollo-generator) ✗ npm test

> open_api_petstore@1.0.0 test
> mocha --require @babel/register --recursive


ReferenceError: factory is not defined
    at Object.<anonymous> (/Users/williamcheng/Code/openapi-generator4/samples/client/petstore/javascript-apollo/test/api/PetApi.spec.js:15:1)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Module._compile (/Users/williamcheng/Code/openapi-generator4/samples/client/petstore/javascript-apollo/node_modules/pirates/lib/index.js:136:24)
    at Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Object.newLoader [as .js] (/Users/williamcheng/Code/openapi-generator4/samples/client/petstore/javascript-apollo/node_modules/pirates/lib/index.js:141:7)
    at Module.load (node:internal/modules/cjs/loader:981:32)

```

joaomlneto commented:
@wing328 I encountered that problem before I deleted `node_modules` (or maybe even regenerating the samples?) and retried `npm install && npm test`, not exactly sure why.

```
rm -rf samples/client/petstore/javascript-*
./run-in-docker.sh /gen/bin/generate-samples.sh bin/configs/javascript-*
(cd samples/client/petstore/javascript-apollo && npm install && npm test)
(cd samples/client/petstore/javascript-es6 && npm install && npm test)
(cd samples/client/petstore/javascript-promise-es6 && npm install && npm test)
```

wing328 commented:
I tested it in the cirlceci but got the same errors:
```
> open_api_petstore@1.0.0 test
> mocha --require @babel/register --recursive


ReferenceError: factory is not defined
    at Object.<anonymous> (/home/circleci/OpenAPITools/openapi-generator/samples/client/petstore/javascript-apollo/test/api/PetApi.spec.js:15:1)
    at Module._compile (node:internal/modules/cjs/loader:1126:14)
    at Module._compile (/home/circleci/OpenAPITools/openapi-generator/samples/client/petstore/javascript-apollo/node_modules/pirates/lib/index.js:136:24)
    at Module._extensions..js (node:internal/modules/cjs/loader:1180:10)
    at Object.newLoader [as .js] (/home/circleci/OpenAPITools/openapi-generator/samples/client/petstore/javascript-apollo/node_modules/pirates/lib/index.js:141:7)
    at Module.load (node:internal/modules/cjs/loader:1004:32)
    at Function.Module._load (node:internal/modules/cjs/loader:839:12)
    at Module.require (node:internal/modules/cjs/loader:1028:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.exports.requireOrImport (/home/circleci/OpenAPITools/openapi-generator/samples/client/petstore/javascript-apollo/node_modules/mocha/lib/esm-utils.js:42:12)
    at Object.exports.loadFilesAsync (/home/circleci/OpenAPITools/openapi-generator/samples/client/petstore/javascript-apollo/node_modules/mocha/lib/esm-utils.js:55:34)
    at singleRun (/home/circleci/OpenAPITools/openapi-generator/samples/client/petstore/javascript-apollo/node_modules/mocha/lib/cli/run-helpers.js:125:3)
    at Object.exports.handler (/home/circleci/OpenAPITools/openapi-generator/samples/client/petstore/javascript-apollo/node_modules/mocha/lib/cli/run.js:362:5)
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:exec (mocha) on project openapi-v3-petstore-javascript-apollo: Command execution failed.: Process exited with an error: 1 (Exit value: 1) -> [Help 1]
[ERROR]
```
https://app.circleci.com/pipelines/github/OpenAPITools/openapi-generator/19176/workflows/78ef7f63-4493-4cd3-8260-82327c9dbc1c/jobs/47447

wing328 commented:
I'm using node v16.13.0. Which version are you using?

joaomlneto commented:
This is odd. I'll take some time to understand what's up with these heisenbugs.

---

> I'm using node v16.13.0. Which version are you using?

I was on node 18 (`v18.8.0`) but downgraded[1] to 16 (`v16.17.0`) and still works just fine with the commands in my [previous comment](https://github.com/OpenAPITools/openapi-generator/pull/13270#issuecomment-1227209884). My NPM version: `8.15.0`

wing328 commented:
No luck even after updating npm to the latest version `8.18.0`

joaomlneto commented:
@wing328 I managed to reproduce both your and my outputs.

The problem is I thought `generate-samples.sh` would delete any old/unused files from the `samples` folder, and regenerate it strictly from the current state of the codebase. Took me a while to realize this, by doing some detective work.

In the end, this is how I reproduced and fixed it:

(1) `./run-in-docker.sh mvn clean package && ./run-in-docker.sh /gen/bin/generate-samples.sh && ./run-in-docker.sh /gen/bin/utils/export_docs_generators.sh`

(2) Running `npm install && npm test` on the javascript samples fails
- `(cd samples/client/petstore/javascript-apollo && npm install && npm test)` ➡️ 💥 `factory is not defined`
- `(cd samples/client/petstore/javascript-es6 && npm install && npm test)` ➡️  `227 passing, 1 failing ❌`
- `(cd samples/client/petstore/javascript-promise-es6 && npm install && npm test)` ➡️ `233 passing ✅`

(3) `rm -rf samples/client/petstore/javascript-* && ./run-in-docker.sh /gen/bin/generate-samples.sh bin/configs/javascript-*`

(4) Running `npm install && npm test` on the regenerated-from-scratch javascript samples now succeeds
- `(cd samples/client/petstore/javascript-apollo && npm install && npm test)` ➡️ 💥 `209 passing ✅`
- `(cd samples/client/petstore/javascript-es6 && npm install && npm test)` ➡️  `209 passing ✅`
- `(cd samples/client/petstore/javascript-promise-es6 && npm install && npm test)` ➡️ `209 passing ✅`

joaomlneto commented on bin/configs/unmaintained/javascript-apollo.yaml:
```diff
@@ -1,5 +1,5 @@
 generatorName: javascript-apollo
-outputDir: samples/client/petstore/javascript-apollo
+outputDir: samples/client/petstore/javascript-apollo-unmaintained
```
In the end there was no config clash with [`bin/configs/javascript-apollo.yaml`](https://github.com/OpenAPITools/openapi-generator/blob/8070b4057f3bd43df3e40117906f09d65d36c015/bin/configs/javascript-apollo.yaml), as `unmaintained` configs are not used by `generate-samples.sh`.

wing328 replied:
I'll remove it later.

wing328 replied:
Done via https://github.com/OpenAPITools/openapi-generator/pull/13297

wing328 commented:
Pull the latest and all tests passed
```
  User
    ✓ should create an instance of User
    ✓ should have the property id (base name: "id")
    ✓ should have the property username (base name: "username")
    ✓ should have the property firstName (base name: "firstName")
    ✓ should have the property lastName (base name: "lastName")
    ✓ should have the property email (base name: "email")
    ✓ should have the property password (base name: "password")
    ✓ should have the property phone (base name: "phone")
    ✓ should have the property userStatus (base name: "userStatus")


  209 passing (405ms)

```

wing328 commented on samples/client/petstore/javascript-promise-es6/test/ApiClientTest.js:
```diff
@@ -1,394 +0,0 @@
-if (typeof module === 'object' && module.exports) {
```
looks like some tests were removed. i'll try to restore them.

wing328 replied:
filed https://github.com/OpenAPITools/openapi-generator/pull/13547
